File: 876-middle-of-the-linked-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
/* https://leetcode.com/problems/middle-of-the-linked-list/ */
function middleNode(head: ListNode | null): ListNode | null {
  let ahead = head;

  while (ahead && ahead.next) {
    ahead = ahead.next.next;
    head = head.next;
  }

  return head;
}

File: 876-middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# https://leetcode.com/problems/middle-of-the-linked-list/
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        ahead = head

        while ahead and ahead.next:
            ahead = ahead.next.next
            head = head.next

        return head
